crates/rpc/src/cairo/ext_py/ser.rs

```diff
@@ -166,74 +148,81 @@ impl<'a> serde::Serialize for DiffElement<'a> {
 ///
 /// In call.py this is read by `def maybe_pending_deployed`.
 #[derive(Debug)]
-pub struct DeployedContractsWrapper<'a>(Option<&'a [DeployedContract]>);
+pub struct DeployedContracts<'a>(Option<&'a StateUpdate>);
 
-impl<'a> From<Option<&'a PendingStateUpdate>> for DeployedContractsWrapper<'a> {
-    fn from(u: Option<&'a PendingStateUpdate>) -> Self {
-        let cs = u.map(|u| u.state_diff.deployed_contracts.as_slice());
-        DeployedContractsWrapper(cs)
+impl<'a> From<Option<&'a StateUpdate>> for DeployedContracts<'a> {
+    fn from(u: Option<&'a StateUpdate>) -> Self {
+        Self(u)
     }
 }
 
-impl<'a> serde::Serialize for DeployedContractsWrapper<'a> {
+impl<'a> serde::Serialize for DeployedContracts<'a> {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: serde::Serializer,
     {
         use serde::ser::SerializeSeq;
-        if let Some(cs) = self.0 {
-            let mut seq = serializer.serialize_seq(Some(cs.len()))?;
-            for contract in cs {
-                seq.serialize_element(&DeployedContractElement(contract))?;
-            }
-            seq.end()
-        } else {
-            let seq = serializer.serialize_seq(Some(0))?;
-            seq.end()
+
+        #[derive(serde::Serialize)]
+        struct DeployedContract {
+            address: ContractAddress,
+            contract_hash: ClassHash,
         }
-    }
-}
 
-struct DeployedContractElement<'a>(&'a DeployedContract);
+        let contracts = self
+            .0
+            .iter()
+            .flat_map(|x| x.contract_updates.iter())
+            // Note that this includes both deployed and replaced classes.
+            .filter_map(|(addr, update)| update.class.as_ref().map(|c| (*addr, c.class_hash())));
```

Comment:
I'm not sure if this is correct or not - does cairo-lang need to know about replaced classes as well? Or only the deployed classes?

The latter is the current behaviour on main but I imagine this just got overlooked when replaced classes came into being?

Reply:
Yeah, in theory unknown classes may just appear pretty much anywhere (we've seen some in `library_call` syscalls). So I think this is correct.